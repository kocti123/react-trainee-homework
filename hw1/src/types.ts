import { AnyAction } from "redux";
import { ThunkAction } from "redux-thunk";

export interface ITask {
  readonly id: number;
  readonly body: string;
  readonly completed: boolean;
  readonly favourite: boolean;
  readonly isEdit: boolean;
  readonly newBody: string;
  readonly creationDate: Date;
}

export type filters = "all" | "completed" | "favourite" | "notCompleted";

export type TasksState = {
  readonly isLoaded: boolean;
  readonly lastId: number;
  readonly tasks: ITask[];
  readonly currentFilter: filters;
  readonly filtredTasks: ITask[];
};

export type State = {
  readonly tasks: TasksState;
};

export type AppAction = ThunkAction<void, State, unknown, AnyAction>;
