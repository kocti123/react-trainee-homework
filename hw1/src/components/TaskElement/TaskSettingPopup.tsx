import { useEffect, useRef } from "react";
import ReactDOM from "react-dom";
import { useAppDispatch } from "../../redux";
import {
  removeFromCompleted,
  markTaskAsCompleted,
  removeFromFavourite,
  markTaskAsFavourite,
  setIsEdit,
} from "../../redux/actions/tasksActions";
import { ITask } from "../../types";
import styles from "./TaskSettingPopup.module.css";

const popupDOM = document.querySelector<HTMLDListElement>("#popup")!;

function TaskSettingPopup({
  onClose,
  onDelete,
  task,
  id,
  position,
}: {
  onClose: () => void;
  onDelete: (task: ITask) => void;
  task: ITask;
  id: number;
  position: { x: number; y: number };
}) {
  const dispatch = useAppDispatch();
  const { completed, favourite, isEdit } = task;
  const popupRef = useRef<HTMLUListElement>(null);

  useEffect(() => {
    if (popupRef.current) {
      popupRef.current.style.top = `${position.y}px`;
      popupRef.current.style.left = `${position.x}px`;
    }
  }, [position, popupRef]);

  function toggleCompleted() {
    if (completed) {
      dispatch(removeFromCompleted(id));
    } else {
      dispatch(markTaskAsCompleted(id));
    }
    onClose();
  }

  function toggleFavourite() {
    if (favourite) {
      dispatch(removeFromFavourite(id));
    } else {
      dispatch(markTaskAsFavourite(id));
    }
    onClose();
  }

  function toggleIsEdit() {
    if (isEdit) {
      dispatch(setIsEdit(id, false));
    } else {
      dispatch(setIsEdit(id, true));
    }
    onClose();
  }

  function removeTaskHandler() {
    onDelete(task);
    onClose();
  }

  return ReactDOM.createPortal(
    <>
      <div className={styles.popupWrapper} onClick={onClose}></div>
      <ul ref={popupRef} className={styles.popup}>
        <li>
          <button onClick={toggleFavourite}>
            {favourite ? "Убрать из избранного" : "В избранное"}
          </button>
        </li>
        <li>
          <button onClick={toggleCompleted}>
            {completed ? "Вернуть в работу" : "Выполненно"}
          </button>
        </li>
        <li>
          <button onClick={toggleIsEdit}>
            {isEdit ? "Отменить редактирование" : "Редактировать"}
          </button>
        </li>
        <li>
          <button onClick={removeTaskHandler}>Удалить</button>
        </li>
      </ul>
    </>,
    popupDOM
  );
}

export default TaskSettingPopup;
