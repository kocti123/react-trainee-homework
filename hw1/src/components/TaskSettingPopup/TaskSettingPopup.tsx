import { useEffect, useRef } from "react";
import ReactDOM from "react-dom";
import { useAppDispatch } from "../../redux";
import {
  removeFromCompleted,
  markTaskAsCompleted,
  removeFromFavourite,
  markTaskAsFavourite,
  setIsEdit,
} from "../../redux/actions/tasksActions";
import { ITask } from "../../types";
import styles from "./TaskSettingPopup.module.css";

function TaskSettingPopup({
  onClose,
  onDelete,
  task,
  position,
}: {
  onClose: () => void;
  onDelete: (task: ITask) => void;
  task: ITask;
  position: { x: number; y: number };
}) {
  const dispatch = useAppDispatch();
  const { completed, favourite, isEdit, id } = task;
  const popupDOM = document.querySelector<HTMLDListElement>("#popup")!;
  const popupRef = useRef<HTMLUListElement>(null);

  useEffect(() => {
    if (popupRef.current) {
      popupRef.current.style.top = `${position.y}px`;
      popupRef.current.style.left = `${position.x}px`;
    }
  }, [position, popupRef]);

  function toggleCompleted() {
    if (completed) {
      dispatch(removeFromCompleted(id));
    } else {
      dispatch(markTaskAsCompleted(id));
    }
    onClose();
  }

  function toggleFavourite() {
    if (favourite) {
      dispatch(removeFromFavourite(id));
    } else {
      dispatch(markTaskAsFavourite(id));
    }
    onClose();
  }

  function toggleIsEdit() {
    if (isEdit) {
      dispatch(setIsEdit(id, false));
    } else {
      dispatch(setIsEdit(id, true));
    }
    onClose();
  }

  function removeTaskHandler() {
    onDelete(task);
    onClose();
  }

  return ReactDOM.createPortal(
    <>
      <div className={styles.popupWrapper} onClick={onClose}></div>
      <ul ref={popupRef} className={styles.popup}>
        <li>
          <button onClick={toggleFavourite} className="toggleFavourite">
            {favourite ? "Убрать из избранного" : "В избранное"}
          </button>
        </li>
        <li>
          <button onClick={toggleCompleted} className="toggleCompletion">
            {completed ? "Вернуть в работу" : "Выполненно"}
          </button>
        </li>
        <li>
          <button onClick={toggleIsEdit} className="toggleIsEdit">
            {isEdit ? "Отменить редактирование" : "Редактировать"}
          </button>
        </li>
        <li>
          <button onClick={removeTaskHandler} className="remove">
            Удалить
          </button>
        </li>
      </ul>
    </>,
    popupDOM
  );
}

export default TaskSettingPopup;
