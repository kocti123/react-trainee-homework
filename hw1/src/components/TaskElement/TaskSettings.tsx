import { useState } from "react";
import { ReactComponent as MenuIcon } from "../../icons/menu.svg";
import { ITask } from "../../types";
import TaskSettingPopup from "./TaskSettingPopup";
import styles from "./TaskSettings.module.css";

function TaskSettingsPopup({
  id,
  task,
  onDelete,
}: {
  id: number;
  task: ITask;
  onDelete: (task: ITask) => void;
}) {
  const [showSetting, setShowSetting] = useState(false);
  const [positionOfElement, setPostionOfElement] = useState({ x: 0, y: 0 });

  function openSetting(event: React.MouseEvent<HTMLButtonElement>) {
    setPostionOfElement({ x: event.pageX, y: event.pageY });
    setShowSetting(true);
  }

  function close() {
    setShowSetting(false);
  }

  return (
    <div className={styles.taskSettings}>
      <button onClick={openSetting} className={styles.menuButton}>
        <MenuIcon />
      </button>
      {showSetting ? (
        <TaskSettingPopup
          onDelete={onDelete}
          onClose={close}
          id={id}
          task={task}
          position={positionOfElement}
        />
      ) : undefined}
    </div>
  );
}

export default TaskSettingsPopup;
