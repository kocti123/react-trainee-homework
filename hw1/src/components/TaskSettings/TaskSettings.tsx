import { useState } from "react";
import { ReactComponent as MenuIcon } from "../../icons/menu.svg";
import { ITask } from "../../types";
import TaskSettingPopup from "../TaskSettingPopup/TaskSettingPopup";
import styles from "./TaskSettings.module.css";

function TaskSettings({
  task,
  onDelete,
}: {
  task: ITask;
  onDelete: (task: ITask) => void;
}) {
  const [showSetting, setShowSetting] = useState(false);
  const [positionOfElement, setPostionOfElement] = useState({ x: 0, y: 0 });

  function openSetting(event: React.MouseEvent<HTMLButtonElement>) {
    setPostionOfElement({ x: event.pageX, y: event.pageY });
    setShowSetting(true);
  }

  function close() {
    setShowSetting(false);
  }

  return (
    <div className={styles.taskSettings}>
      <button onClick={openSetting} className={styles.menuButton}>
        <MenuIcon />
      </button>
      {showSetting ? (
        <TaskSettingPopup
          onDelete={onDelete}
          onClose={close}
          task={task}
          position={positionOfElement}
        />
      ) : undefined}
    </div>
  );
}

export default TaskSettings;
