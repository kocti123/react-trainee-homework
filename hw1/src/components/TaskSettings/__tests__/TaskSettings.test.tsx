import { shallow, ShallowWrapper } from "enzyme";
import { wrap } from "module";
import { ITask } from "../../../types";
import TaskSettingPopup from "../../TaskSettingPopup/TaskSettingPopup";
import TaskSettings from "../TaskSettings";

describe("TaskSetting", () => {
  let wrapper: ShallowWrapper;
  const props: {
    onDelete: any;
    task: ITask;
  } = {
    onDelete: jest.fn(),
    task: {
      id: 1,
      body: "test",
      completed: false,
      favourite: false,
      isEdit: false,
      newBody: "",
      creationDate: new Date(1635768398831),
    },
  };

  beforeEach(() => {
    wrapper = shallow(<TaskSettings {...props} />);
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should open popupSettings when click on button", () => {
    wrapper.find("button").simulate("click", { pageX: 100, pageY: 100 });

    expect(
      wrapper.find("TaskSettingPopup").getElement().props.position
    ).toEqual({ x: 100, y: 100 });
  });

  it("should stop showing popup after onClose call", () => {
    wrapper.find("button").simulate("click", { pageX: 100, pageY: 100 });

    expect(wrapper.children()).toHaveLength(2);

    wrapper.find("TaskSettingPopup").getElement().props.onClose();

    expect(wrapper.children()).toHaveLength(1);
  });
});
