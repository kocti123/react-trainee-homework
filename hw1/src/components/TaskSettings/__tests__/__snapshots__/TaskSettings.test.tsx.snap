// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskSetting should match snapshot 1`] = `ShallowWrapper {}`;
